### PR TITLE
[ADD] contract_invoicing_grouping_criteria

### PR DESCRIPTION
**Pull request by paradoxxxzero**
This module allows to group invoicing of several contracts and to use custom criteria for grouping contracts to be invoiced.

This module mimics the sale.order grouping behavior of the sale module and adds the sale_order_invoicing_grouping_criteria's custom criteria mechanism.

### CODE REVIEW

**Comment by paradoxxxzero** [comment]
@mourad-ehm @hparfr

File: contract_invoicing_grouping_criteria/models/contract.py [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+        # if x is a record, return its id to get a consistent sorting
+        return x.id if isinstance(x, models.BaseModel) else x
+
+    def get_with_or(x, key):
+        """Get key from x. Handles keys containing ||"""
+        for k in key.split("||"):
+            val = getattr(x, k)
+            if val:
+                return val
+        return val
```
Comment (mourad-ehm) [inline contract_invoicing_grouping_criteria/models/contract.py:19]: val must be defined before the for loop if key is empty we can have a bug.

Comment (paradoxxxzero) [reply contract_invoicing_grouping_criteria/models/contract.py:19]: Split with an argument always returns a non empty list
(cf: https://docs.python.org/3.3/library/stdtypes.html#str.split)

Comment (mourad-ehm) [reply contract_invoicing_grouping_criteria/models/contract.py:19]: in this case the last return is useless ?

Comment (paradoxxxzero) [reply contract_invoicing_grouping_criteria/models/contract.py:19]: Not very useful indeed but this does not matter much

**Pull request closed without merging**